Pass core WebAssembly spec tests on ARM64 with single-pass WebAssembly baseline JIT

ddegazio opened the pull request:
#### 46375fbc5a6710ec3e2fd10848009e31b307b741
<pre>
Pass core WebAssembly spec tests on ARM64 with single-pass WebAssembly baseline JIT
<a href="https://bugs.webkit.org/show_bug.cgi?id=250052">https://bugs.webkit.org/show_bug.cgi?id=250052</a>
rdar://105421077

Reviewed by Yusuke Suzuki.

Adds a new single-pass baseline JIT implementation for WebAssembly, the single-pass
BBQ JIT. The main goal of this new compiler is to compile substantially faster than
the existing Air tier, while offering comparable levels of generated code performance.
This should help us start up considerably faster on large WASM programs.

This is the initial commit of this feature, and the new baseline is only partially
complete. Currently, it should pass all the core spec tests on ARM64, but there are
some remaining issues on other platforms, some failures on the full stress tests, and
missing support for some key features like SIMD. As such, it&apos;s being landed disabled
by default, behind the new &apos;--useSinglePassBBQJIT&apos; flag.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::getGlobal):
(JSC::Wasm::parseAndCompileAirImpl):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::parseAndCompileB3):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.h: Added.
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::dumpDisassembly):
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
(JSC::Wasm::BBQPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::typeOfLocal const):

Canonical link: <a href="https://commits.webkit.org/260260@main">https://commits.webkit.org/260260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/986cd6f609f7e7f2c0167cc45e0bfe1b72d28f2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107650 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40545 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116202 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18123 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8019 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99821 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13682 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96879 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41352 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95597 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28509 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83195 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96954 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9688 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29858 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96364 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7715 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10363 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6751 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15839 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49443 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105235 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7092 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11923 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26093 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Uzlopak commented on Source/JavaScriptCore/wasm/WasmBBQIRGenerator.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2019 Apple Inc. All rights reserved.
```
![image](https://media.tenor.com/UxhRnOP359QAAAAC/wtf-time.gif)

saambarati commented on Source/JavaScriptCore/wasm/WasmBBQIRGenerator.cpp:
```diff
@@ -0,0 +1,3383 @@
+        }
+    };
+
+    #define BLOCK(...) __VA_ARGS__
```
Why?

ddegazio replied:
It's meant to shield any commas in the parameters of `EMIT_UNARY` and `EMIT_BINARY` from being interpreted as macro parameter separators. It's probably unnecessary for most opcode implementations, since most of the time commas only occur within (function) parameter lists anyway, but it prevents weirdness with multiple template parameters (i.e. `emitModOrDiv<int32_t, false>(...)`) or comma expressions if they do occur. 

It is worth evaluating if there's any case where this is truly necessary though, as opposed to something simpler like a curly-braced block. I'll give it a look! 👀

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.h:
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (C) 2019-2022 Apple Inc. All rights reserved.
```
2023

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    using DataLabel32 = MacroAssembler::DataLabel32;
+
+    // Functions can have up to 1000000 instructions, so 32 bits is a sensible maximum number of stack items or locals.
+    using LocalOrTempIndex = uint32_t;
```
static_assert?

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+            bind(local, location);
+        }
+        emitMove(value, location);
+        kill(value);
```
Let's rename this

ddegazio replied:
Renamed in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -2615,7 +2622,7 @@ FOR_EACH_WASM_MEMORY_STORE_OP(CREATE_CASE)
     case Drop: {
         WASM_PARSER_FAIL_IF(!m_expressionStack.size(), "can't drop on empty stack");
-        m_expressionStack.takeLast();
+        WASM_TRY_ADD_TO_CONTEXT(drop(m_expressionStack.takeLast()));
```
For callback, name it as addDrop

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    int m_frameSize = 0;
+    int m_maxCalleeStackSize = 0;
+    int m_localStorage = 0; // Stack offset pointing to the local with the lowest address.
+    constexpr static int tempSlotSize = 8; // Size of the stack slot for a stack temporary.
```
In WebKit coding style, we use `int m_frameSize { 0 };` instead of ` = 0;`.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    Vector<uint32_t, 4> m_arguments;
+    ControlData m_topLevel;
+
+    Vector<RegisterBinding, 32> m_gprBindings, m_fprBindings; // Stack of both local and stack bindings.
```
In WebKit coding style, we define them separately,
```
Vector<RegisterBinding, 32> m_gprBindings;
Vector<RegisterBinding, 32> m_fprBindings;
```

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    Vector<RegisterBinding, 32> m_gprBindings, m_fprBindings; // Stack of both local and stack bindings.
+    RegisterSet m_gprSet, m_fprSet; // Sets tracking whether registers are bound or free.
+    RegisterSet m_validGPRs, m_validFPRs; // These contain the original register sets used in m_gprSet and m_fprSet.
+    Vector<Location, 8> m_locals, m_temps; // Vectors mapping local and temp indices to binding indices.
```
Ditto.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    FPRegisterID m_scratchFPR;
+    RegisterID m_dataScratchGPR; // Used specifically as a temporary for complex moves.
+
+    RegisterID m_wasmContextInstanceGPR, m_memoryBaseGPR, m_boundsCheckingSizeGPR;
```
I think we no longer need to have this. `GPRInfo::` knows them. And we no longer support TLS version.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+    // end of the function.
+    struct PrologueInfo {
+        RegisterSet usedCalleeSavedGPRs, usedCalleeSavedFPRs;
+        bool hasExceptionHandlers, mayClobberWasmContextInstance;
```
Ditto.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+        if (m_info.isImportedFunctionFromFunctionIndexSpace(functionIndex)) {
+            m_jit.move(TrustedImmPtr(Instance::offsetOfImportFunctionStub(functionIndex)), m_scratchGPR);
+            m_jit.addPtr(m_wasmContextInstanceGPR, m_scratchGPR);
+            m_jit.loadPtr(Address(m_scratchGPR, 0), m_scratchGPR);
```
If offset is 0, you do not need to put it. `Address(m_scratchGPR)` works.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+        // Materialize address of native function and call register
+        void* taggedFunctionPtr = tagCFunctionPtr<void*, OperationPtrTag>(function);
+        m_jit.move(TrustedImmPtr(bitwise_cast<uintptr_t>(taggedFunctionPtr)), m_scratchGPR);
+        m_jit.call(m_scratchGPR, OperationPtrTag);
```
Before CCall, you need to do `prepareWasmCallOperation`

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+            callerSaveFprs.remove(reg);
+        });
+
+        m_scratchGPR = GPRInfo::nonPreservedNonArgumentGPR0;
```
Let's define m_scratchGPR as a `static constexpr GPRReg scratchGPR = GPRInfo::nonPreservedNonArgumentGPR0;` for this class since it does not change.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+        });
+
+        m_scratchGPR = GPRInfo::nonPreservedNonArgumentGPR0;
+        m_dataScratchGPR = GPRInfo::wasmScratchGPR0;
```
Ditto.

ddegazio replied:
Resolved in latest push.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -0,0 +1,5870 @@
+
+        m_boundsCheckingSizeGPR = GPRInfo::wasmBoundsCheckingSizeRegister;
+        if (mode == MemoryMode::BoundsChecking)
+            gprSetBuilder.remove(m_boundsCheckingSizeGPR);
```
Let's just use GPRInfo::wasmBoundsCheckingSizeRegister etc. directly.

ddegazio replied:
Resolved in latest push.

Constellation reviewed (approved):
r=me

ddegazio committed:
Pass core WebAssembly spec tests on ARM64 with single-pass WebAssembly baseline JIT
https://bugs.webkit.org/show_bug.cgi?id=250052
rdar://105421077

Reviewed by Yusuke Suzuki.

Adds a new single-pass baseline JIT implementation for WebAssembly, the single-pass
BBQ JIT. The main goal of this new compiler is to compile substantially faster than
the existing Air tier, while offering comparable levels of generated code performance.
This should help us start up considerably faster on large WASM programs.

This is the initial commit of this feature, and the new baseline is only partially
complete. Currently, it should pass all the core spec tests on ARM64, but there are
some remaining issues on other platforms, some failures on the full stress tests, and
missing support for some key features like SIMD. As such, it's being landed disabled
by default, behind the new '--useSinglePassBBQJIT' flag.

* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::getGlobal):
(JSC::Wasm::parseAndCompileAirImpl):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::parseAndCompileB3):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.h: Added.
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::dumpDisassembly):
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::compileFunction):
(JSC::Wasm::BBQPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser::typeOfLocal const):

Canonical link: https://commits.webkit.org/260260@main
